bindings/rust/bench/src/openssl.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    harness::{CipherSuite, ConnectedBuffer, CryptoConfig, ECGroup, Mode, TlsBenchHarness},
+    CA_CERT_PATH, SERVER_CERT_CHAIN_PATH, SERVER_KEY_PATH,
+};
+use openssl::ssl::{
+    ErrorCode, Ssl, SslAcceptor, SslConnector, SslFiletype, SslMethod, SslStream, SslVersion,
+};
+use std::error::Error;
+
+pub struct OpenSslHarness {
+    client_conn: SslStream<ConnectedBuffer>,
+    server_conn: SslStream<ConnectedBuffer>,
+}
+
+impl OpenSslHarness {
+    /// Process handshake for one connection, treating blocking errors as `Ok`
+    fn handshake_conn(&mut self, mode: Mode) -> Result<(), Box<dyn Error>> {
+        match match mode {
+            Mode::Client => self.client_conn.connect(),
+            Mode::Server => self.server_conn.accept(),
+        } {
+            Ok(_) => Ok(()),
+            Err(err) => {
+                if err.code() != ErrorCode::WANT_READ {
+                    Err(err.into())
+                } else {
+                    Ok(())
+                }
+            }
+        }
+    }
+}
+
+impl TlsBenchHarness for OpenSslHarness {
+    fn new(crypto_config: &CryptoConfig) -> Result<Self, Box<dyn Error>> {
+        let client_buf = ConnectedBuffer::new();
+        let server_buf = client_buf.clone_inverse();
+
+        let cipher_suite = match crypto_config.cipher_suite {
+            CipherSuite::AES_128_GCM_SHA256 => "TLS_AES_128_GCM_SHA256",
+            CipherSuite::AES_256_GCM_SHA384 => "TLS_AES_256_GCM_SHA384",
+        };
+
+        let ec_key = match crypto_config.ec_group {
+            ECGroup::SECP256R1 => "P-256",
+            ECGroup::X25519 => "X25519",
+        };
+
+        let mut client_builder = SslConnector::builder(SslMethod::tls())?;
+        client_builder.set_ca_file(CA_CERT_PATH)?;
+        client_builder.set_min_proto_version(Some(SslVersion::TLS1_3))?;
+        client_builder.set_ciphersuites(cipher_suite)?;
+        client_builder.set_groups_list(ec_key)?;
+
+        let mut server_builder = SslAcceptor::mozilla_modern_v5(SslMethod::tls())?;
+        server_builder.set_certificate_chain_file(SERVER_CERT_CHAIN_PATH)?;
```

Comment:
Looks like all the config for the builders is the same except CA configuration. Should we extract this out into another method?

Reply:
I think extracted out looks only marginally better, but with more features in the future there should be more of a clear distinction.